### PR TITLE
Swift 4

### PR DESCRIPTION
**Pull request by lfarah**
WIP branch of Swift 4. Everybody is invited to help it out.

Checklist
- [x] Building most extensions
- [x] Passing most of Tests
- [ ] Building 100%
- [ ] Passing all Tests

1. We need to ensure all the extensions that were modified are being currently tested
2. We need to ensure all tests work
3. No extension should be removed unless there's a good reason (there's a better way to do it w/ Apple's methods)

## Checklist

- [ ] New Extension
- [ ] New Test
- [x] Changed more than one extension, but all changes are related
- [ ] Trivial change (doesn't require changelog)

Related to #456

### CODE REVIEW

**Comment by Khalian** [comment]
I presume we will be putting out a swift 4.0 branch for now and not merge into master? (Until of course swift 4 is stable and not in beta)?

File: EZSwiftExtensionsTests/ArrayTests.swift [hunk 1]
```diff
@@ -225,9 +225,10 @@ class ArrayTests: XCTestCase {
         XCTAssertEqual(totalIndexes, 10)
         XCTAssertEqual(totalNumbers, 20)
 
-        emptyArray.forEachEnumerated { _ in XCTFail() }
+        emptyArray.forEachEnumerated { _,_  in XCTFail() }
         let copyArray = someArray
-        copyArray.forEachEnumerated { XCTAssertTrue(someArray[$0.0] == $0.1) }
+//        copyArray.forEachEnumerated { XCTAssertTrue(someArray[$0.0] == $0.1) }
+        XCTFail()
```
Comment (Khalian) [inline EZSwiftExtensionsTests/ArrayTests.swift]: Lets put TODOs here and everywhere else we have interjected a deliberate Fail.

**Comment by lfarah** [comment]
As in today, Xcode 9 is GM so it is stable and not in beta anymore

**Comment by lfarah** [comment]
But this branch needs some work before we merge into master

File: EZSwiftExtensionsTests/StringTests.swift [hunk 2]
```diff
@@ -371,16 +372,16 @@ class StringTests: XCTestCase {
         let testString = "meh"
         let testString2 = "✅"
 
-        let boldResult = NSAttributedString(string: testString, attributes: [NSFontAttributeName: UIFont.boldSystemFont(ofSize: UIFont.systemFontSize)])
-        let boldResult2 = NSAttributedString(string: testString2, attributes: [NSFontAttributeName: UIFont.boldSystemFont(ofSize: UIFont.systemFontSize)])
-        let underlineResult = NSAttributedString(string: testString, attributes: [NSUnderlineStyleAttributeName: NSUnderlineStyle.styleSingle.rawValue])
-        let underlineResult2 = NSAttributedString(string: testString2, attributes: [NSUnderlineStyleAttributeName: NSUnderlineStyle.styleSingle.rawValue])
+            let boldResult = NSAttributedString(string: testString, attributes: [NSAttributedStringKey.font: UIFont.boldSystemFont(ofSize: UIFont.systemFontSize)])
```
Comment (Khalian) [inline EZSwiftExtensionsTests/StringTests.swift]: Is this a new lint thing? Or did you interject these spaces accidentally?

File: Sources/UIViewExtensions.swift [hunk 3]
```diff
@@ -445,12 +445,12 @@ extension UIView {
     }
 
     //EZSE: Reverse pop, good for button animations
-    public func reversePop() {
-        setScale(x: 0.9, y: 0.9)
-        UIView.animate(withDuration: 0.05, delay: 0, options: UIViewAnimationOptions.allowUserInteraction, animations: { [weak self] _ in
-            self?.setScale(x: 1, y: 1)
-        })
-    }
+//    public func reversePop() {
```
Comment (Khalian) [inline Sources/UIViewExtensions.swift]: Put a TODO here to fix this back.

File: Sources/UIViewExtensions.swift [hunk 4]
```diff
@@ -609,7 +609,7 @@ extension UIView {
 
 // MARK: Fade Extensions
 
-private let UIViewDefaultFadeDuration: TimeInterval = 0.4
+public let UIViewDefaultFadeDuration: TimeInterval = 0.4
```
Comment (Khalian) [inline Sources/UIViewExtensions.swift]: Is this really a swift 4.0 thing?

File: Sources/StringExtensions.swift [hunk 5]
```diff
@@ -393,24 +393,24 @@ extension String {
...
-        var attrib: [String: AnyObject] = [NSFontAttributeName: font]
-        if lineBreakMode != nil {
-            let paragraphStyle = NSMutableParagraphStyle()
-            paragraphStyle.lineBreakMode = lineBreakMode!
-            attrib.updateValue(paragraphStyle, forKey: NSParagraphStyleAttributeName)
-        }
-        let size = CGSize(width: width, height: CGFloat(Double.greatestFiniteMagnitude))
-        return ceil((self as NSString).boundingRect(with: size, options: NSStringDrawingOptions.usesLineFragmentOrigin, attributes:attrib, context: nil).height)
-    }
+//    public func height(_ width: CGFloat, font: UIFont, lineBreakMode: NSLineBreakMode?) -> CGFloat {
```
Comment (Khalian) [inline Sources/StringExtensions.swift]: Put TODO here to uncomment it later on.

File: Sources/StringExtensions.swift [hunk 6]
```diff
@@ -257,22 +257,22 @@ extension String {
...
-        let from16 = utf16.startIndex.advanced(by: nsRange.location)
-        let to16 = from16.advanced(by: nsRange.length)
-        if let from = String.Index(from16, within: self),
-            let to = String.Index(to16, within: self) {
-            return from ..< to
-        }
-        return nil
-    }
+//    internal func rangeFromNSRange(_ nsRange: NSRange) -> Range<String.Index>? {
+//        let from16 = utf16.startIndex + nsRange.location
```
Comment (Khalian) [inline Sources/StringExtensions.swift]: Put TODO here and the next method that they will be uncommented later.

File: Sources/DictionaryExtensions.swift [hunk 7]
```diff
@@ -19,7 +19,9 @@ extension Dictionary {
     public func union(_ dictionaries: Dictionary...) -> Dictionary {
         var result = self
         dictionaries.forEach { (dictionary) -> Void in
-            dictionary.forEach { (key, value) -> Void in
+            dictionary.forEach { (arg) -> Void in
+                
+                let (key, value) = arg
```
Comment (Khalian) [inline Sources/DictionaryExtensions.swift]: Is this necessary in swift 4 syntax? IMO the earlier version looked cleaner.

File: Sources/ArrayExtensions.swift [hunk 8]
```diff
@@ -220,7 +220,7 @@ extension Array where Element: Hashable {
 extension Collection where Indices.Iterator.Element == Index {
     
     /// Returns the element at the specified index if it is within bounds, otherwise nil.
-    public subscript (safe index: Index) -> Generator.Element? {
+    public subscript (safe index: Index) -> Iterator.Element? {
```
Comment (Khalian) [inline Sources/ArrayExtensions.swift]: Whats the difference?

**Comment by lfarah** [comment]
Alright, I made a PR based on (@Steven-Cheung's PR](https://github.com/goktugyil/EZSwiftExtensions/pull/470)

**Comment by Khalian** [comment]
Already done.

**Pull request closed without merging**